requests-for-comments/2021-10-05-group-analytics.md

```diff
@@ -0,0 +1,218 @@
+
+From speaking with clients, their internal amount of different “group types” (e.g. `organization`) is limited, though the number of groups under a group type is unlimited (e.g. Facebook if it was an organization).
+
+This constraint allows us to implement things slightly more efficiently and have a nicer UX compared to unbounded amount of groups.
+
+### 4. Feature flags in groups should be consistent
+
+For the users who are using feature flags, they want to ensure a consistent experience within a group during rollout.
+
+This is to avoid situations where some users see a new feature and some do not.
+
+This also factors into their later analysis - for example, they’d like to treat two different “organizations” both as a single datapoint even if one has 1500 users and the other has 1.
+
+## Proposal
+
+### Groups are event-level
+
+Users can define up to N (5?) group types. For us these would be `instance` , `project` and `organization`.
+
+Each event we ingest can have one group per group type associated. Group types are created when events with that type are captured.
+
+We will have N new columns in the `events` table, containing the group ids - first one for the first group type of the project, second for the second one etc.
+
+### Group properties API
+
+We will expose a new API for creating and adding properties groups. This API can be exposed on the client level, e.g.
+
+```python
+    import posthog
+
+    posthog.groups.add_properties('organization', 'org:5', { name: 'Swedish Bakery', subscription_level: 'gold' })
+```
+
+There will not be any functionality for merging two groups, as that behavior is user-specific.
+
+### Capture API
+
+Capturing events in client libraries will allow users to pass group metadata alongside with an event.
+
+Here’s how it would look like in python:
+
+```python
+    import posthog
+
+    posthog.capture(
+      user_id,
+      'purchase'
+      { plan: 'silver' },
+      groups={
+        'organization': 'org:6'
+      }
+    )
+```
+
+`groups` metadata will be sent to us under event properties under new the `$groups` key.
+
+Our data ingestion in plugin-server will perform the neccessary mappings from group ids to column indexes or other keys as needed.
+
+
+### In-App: filtering by group properties in insights
+
+In-app, users will be able to do analytics by groups and filter by them.
+
+To give an example:
+
+- Throughout insights, we’ll allow filtering insights with group properties. This will be implemented as a new tab per group type in the filters modal.
+- In trends, users can also select “unique organizations” after first capturing events for an “organization”
+- In funnels, users can select to analyze by a group type (e.g. “instance”). Different users within the group can have completed different steps of the funnel.
+- In retention, users can select “organization” over “user” for the unit to analyze retention by
+- And so on…
+
+
+### In-App: profile pages for groups
```

Comment:
I suggest involving @clarkus very early on. Probably worth giving this more thought into what we want to expose to users, perhaps there's a simpler way to display this all together. We would also probably want to add some groups information into the Person page.

Reply:
Will do once there's something to iterate! :)